app/controllers/concerns/authenticates_with_two_factor.rb

```diff
@@ -53,16 +56,43 @@ def authenticate_with_two_factor
 
   private def authenticate_with_two_factor_via_otp(user)
     if valid_otp_attempt?(user) || valid_backup_code_attempt?(user)
-      # Remove any lingering user data from login
-      session.delete(:otp_user_id)
-
-      user.save!
-      sign_in(user, message: :two_factor_authenticated, event: :authentication)
+      # add 2fa device if true
+      add_2fa_device(user, user_params[:device_name]) if user_params[:remember_device] == 'true' && bypass_2fa_enabled?
+      two_factor_successful(user)
     else
       user.increment_failed_attempts
       Rails.logger.info("Failed Login: user=#{user.email} ip=#{request.remote_ip} method=OTP")
       flash.now[:alert] = _('Invalid two-factor code.')
       prompt_for_two_factor(user)
     end
   end
+
+  private def add_2fa_device(user, name)
+    uuid = SecureRandom.uuid
+
+    # set cookie and add to two_factors_memorized_devices list
+    cookies.encrypted[:memorized_device] = { value: uuid, expires: GrdaWarehouse::Config.get(:bypass_2fa_duration).days.from_now }
+    user.two_factors_memorized_devices.create!(
+      uuid: uuid,
+      name: name,
+      session_id: session.id,
+      log_in_ip: request.remote_ip,
+    )
+  end
+
+  private def using_memorized_device?(user)
+    cookie_uuid = cookies.encrypted[:memorized_device]
+    return false unless cookie_uuid
+
+    # find if cookie exist in active
+    user.two_factors_memorized_devices.active.exists?(uuid: cookie_uuid)
```

Comment:
Nice! I'd not seen this use of `exists?`.  Usually we have `scope.exists?` which isn't quite as nice as this for looking up an individual record.